### ISSUE TITLE
categorical parameters in config_space (in smac)

### ISSUE THREAD

**Issue by hamid-arjmandi** [automl/ConfigSpace#139]
Apparently, for each run of smac sequence of the ordinal and categorical parameters varies. 
that you can see by printing "config_space.get_hyperparameters()". That leads to non-deterministic conversion of configs from an array **given by the user** using  'Configuration(self.scenario.cs, vector=vector)'. 
It can be easily fixed if the choices of categorical hyperparameters are sorted uniquely (for example alphabetically) once they are created. The same problem exists for ordinal hyperparameters.
we are using ConfigSpace-0.4.11.

Example:
if you have a **constant** array with the same length as config space parameters, the same configuration will not be generated every time (change only in categorical/ordinal parameters). 

        x = [1., 2., 0.8, 0.2, 2., 0.1, 1.8, 0.31176448, 0.42509928, 0.05, 0.0990991, 0.05]
        print(Configuration(self.scenario.cs, vector=x))
[cs.txt](https://github.com/automl/ConfigSpace/files/3967753/cs.txt)

**Comment by hamid-arjmandi** [comment]
The Configuration function is imported from smac.configspace

**Comment by mfeurer** [comment]
Thanks a lot for repording this issue @hamid-arjmandi 

I produced the following snippet which according to your report should trigger the error:
```
from ConfigSpace.hyperparameters import CategoricalHyperparameter, UniformIntegerHyperparameter, \
    UniformFloatHyperparameter
from ConfigSpace import ConfigurationSpace, Configuration

var1 = CategoricalHyperparameter('var1', ['ch1', 'ch2'], 'ch2')
var2 = CategoricalHyperparameter('var2', ['ch3', 'ch4', 'ch5'], 'ch5')
var3 = UniformIntegerHyperparameter('var3', 0, 16, 1)
var4 = UniformFloatHyperparameter('var4', 1e-07, 2.0, 0.1)
var5 = UniformFloatHyperparameter('var5', 0.0, 2.0, 0.1)
var6 = UniformFloatHyperparameter('var6', 0.001, 1.0, 0.1)
var7 = UniformFloatHyperparameter('var7', 0.0, 2.0, 0.1)
var8 = CategoricalHyperparameter('var8', ['ch6', 'ch7', 'ch8', 'ch9'], 'ch8')
var9 = UniformFloatHyperparameter('var9', 0.0, 1.0, 0.0)
var10 = CategoricalHyperparameter('var10', ['ch10', 'ch11', 'ch12'], 'ch10')
var11 = CategoricalHyperparameter('var11', ['false', 'true'], 'false')
var12 = CategoricalHyperparameter('var12', ['false', 'true'], 'true')

cs = ConfigurationSpace()
cs.add_hyperparameters([var1, var2, var3, var4, var5, var6, var7, var8, var9, var10, var11, var12])

x = [1., 2., 0.8, 0.2, 2., 0.1, 1.8, 0.31176448, 0.42509928, 0.05, 0.0990991, 0.05]
print(Configuration(cs, vector=x))
```
However, the snippet is deterministic for me. Would it be possible that you provide me with code to generate the configuration space (at least the parts where the categorical hyperparameters are created)?

**Comment by hamid-arjmandi** [comment]
Thanks a lot for your reply. 
The problem was coming from the our wrong implementation:

  var1 = CategoricalHyperparameter(name='var1', choices={'ch1',  'ch2'}, default_value='ch1'))

The problem is in a strange way solved when we used your suggestion:
var1 = CategoricalHyperparameter('var1', ['ch1', 'ch2'], 'ch2')

Therefore I announce the issue closed. 
Thanks a lot for your help and time.

Best regards,
Hamid

**Comment by mfeurer** [comment]
Thanks for the answer. The difference between the two implementations is the failing one uses a set to store the different choices while the other uses a list for the different choices. Sets are by definition unsorted and can therefore have a different order in each invocation of the source code. Lists on the other hand are always ordered, and therefore, the program behaves the same across invocations. I hope this helps.

**Comment by mfeurer** [comment]
Reopening this as we can easily catch this error to avoid such issues.

**Comment by mfeurer** [comment]
Fixed in #152.
